bgpd/bgp_zebra.c

```diff
@@ -2316,6 +2316,102 @@ static int iptable_notify_owner(ZAPI_CALLBACK_ARGS)
 	return 0;
 }
 
+/* Process route notification messages from RIB */
+static int bgp_zebra_route_notify_owner(int command, struct zclient *zclient,
+					zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct prefix p;
+	enum zapi_route_notify_owner note;
+	uint32_t table_id;
+	char buf[PREFIX_STRLEN];
+	afi_t afi;
+	safi_t safi;
+	struct bgp_node *rn;
+	struct bgp *bgp;
+	struct bgp_path_info *pi, *new_select;
+
+	if (!zapi_route_notify_decode(zclient->ibuf, &p, &table_id, &note,
+				      &afi, &safi)) {
+		zlog_err("%s : error in msg decode", __PRETTY_FUNCTION__);
+		return -1;
+	}
+
+	/* Get the bgp instance */
+	bgp = bgp_lookup_by_vrf_id(vrf_id);
+	if (!bgp) {
+		flog_err(EC_BGP_INVALID_BGP_INSTANCE,
+				"%s : bgp instance not found vrf %d",
+				__PRETTY_FUNCTION__, vrf_id);
+		return -1;
+	}
+
+	if ((afi < AFI_IP) || (afi >= AFI_MAX) ||
+	    (safi < SAFI_UNICAST) || (safi >= SAFI_MAX)) {
+		flog_err(EC_BGP_INVALID_ROUTE,
+			 "%s : invalid afi %d, safi %d", __PRETTY_FUNCTION__,
+			 afi, safi);
+	}
+
+	if (BGP_DEBUG(zebra, ZEBRA))
+		prefix2str(&p, buf, sizeof(buf));
+
+	/* Find the bgp route node */
+	rn = bgp_afi_node_lookup(bgp->rib[afi][safi], afi, safi, &p,
+				 &bgp->vrf_prd);
+	if (!rn) {
+		flog_err(EC_BGP_INVALID_ROUTE,
+			 "%s : route %s not found afi %d, safi %d",
+			 __PRETTY_FUNCTION__, buf, afi, safi);
+		return -1;
+	}
+	bgp_unlock_node(rn);
+
+	switch (note) {
+	case ZAPI_ROUTE_FAIL_INSTALL:
+		UNSET_FLAG(rn->flags, BGP_NODE_FIB_INSTALLED);
+		flog_err(EC_BGP_FIB_INSTALL_ERROR,
+		    "[%s] Route install failure for afi %d, safi %d",
+			buf, afi, safi);
+		break;
+	case ZAPI_ROUTE_INSTALLED:
+		new_select = NULL;
+		/* Clear the flags so that route can be processed */
+		if (CHECK_FLAG(rn->flags, BGP_NODE_FIB_INSTALL_PENDING)) {
+			UNSET_FLAG(rn->flags, BGP_NODE_FIB_INSTALL_PENDING);
+			SET_FLAG(rn->flags, BGP_NODE_FIB_INSTALLED);
+			if (BGP_DEBUG(zebra, ZEBRA))
+				zlog_debug("route %s : BGP_NODE_FIB_INSTALLED",
+					   buf);
+			/* Find the best route */
+			for (pi = rn->info; pi; pi = pi->next) {
+				/* Process aggregate route */
+				bgp_aggregate_increment(bgp, &p, pi, afi, safi);
+				if (CHECK_FLAG(pi->flags, BGP_PATH_SELECTED))
+					new_select = pi;
+			}
+			/* Advertise the route */
+			if (new_select)
+				group_announce_route(bgp, afi, safi,
+						     rn, new_select);
+			else {
+				flog_err(EC_BGP_INVALID_ROUTE,
+					 "selected route %s not found", buf);
+				return -1;
+			}
+		}
+		break;
+	case ZAPI_ROUTE_REMOVED:
+		/* Route deleted from dataplane */
+		UNSET_FLAG(rn->flags, BGP_NODE_FIB_INSTALLED);
+		break;
+	default:
```

Comment:
Remove default and handle all switch cases

Reply:
@kssoman can you please address this comment?

Reply:
removed the default condition